### PR TITLE
Add pyodide-js library (alternative to pyodide.js and webworker.js)

### PR DESCRIPTION
**Pull request by gzuidhof**
I took the liberty to modernize the `src/pyodide.js` script, I was hoping to contribute this back. This addresses the following pain points I had integrating Pyodide into [Starboard](https://starboard.gg/) [Notebook](https://github.com/gzuidhof/starboard-notebook):

* There is no way to embed `pyodide.js` into your bundle without copy pasting it into your own project and editing it yourself. There are a few independent projects which each have taken pyodide.js and altered it slightly: their update path will be painful if any meaningful changes are made.
* There was no (clean) way to import `pyodide.js` without immediately also loading the pyodide runtime.
* It could use some love: a lot of `var` could be replaced with `const` and `let`, and many promises could be replaced with async functions using `await`.
* There was no way to turn off `console.log` or `console.error` when loading packages, even when specifying your own callback.
* The CDN URL was string-replaced at build time.
* `pyodide.js` was a very large script which was hard to follow. I didn't go too far with large refactorings - it can definetely have another pass.
* There was a fair bit of variable shadowing (`resolve`).

This PR introduces a new `pyodide-js` package which can be published to NPM. 

Changes:
* There are now 4 entrypoints: 
  * `index.mjs` and `index.cjs` for if one wants to bundle pyodide themselves (ES Module & CommonJS)
  *  `webworker.js` and `browser.js` which should be functionallity equivalent to the existing scripts, we can replace `pyodide.js` by `browser.js` and nothing should change from a user's perspective. Here's the [built version](https://gist.github.com/gzuidhof/c29bec0ae55ea52a6780399b0d526cc9) of `browser.js` as the time of creating this PR.
  * Type definitions are also bundled.
* Rollup is used to create the bundles, and type annotations were added with Typescript.
* There is now a `PyodideLoader` class that keeps some of the state, it can be used directly by those choosing to bundle Pyodide-JS themselves.
* The naming of variables was normalized (there was a mixture of `PascalCase`, `camelCase` and `snake_case` before).
* `async` functions and `await` are used throughout, simplifying the code a lot (especially `preloadWasm()`).
* Split helper functions into a separate file `util.ts`.
* Any `iodide`-specific code was removed.
* `pyodideArtifactsUrl` is now preferred to `languagePluginUrl`, `languagePluginUrl` still works as a fallback (the name doesn't make sense outside of iodide's context).
* If you pass in a message or error callback to `loadPackage`, it no longer logs to console by itself.
* The license of this package is MIT at the moment - but I guess it has to be MPL to be merged into here.

There are definitely still additional improvements to be made, especially to give more control to the user, but I understand this PR is already a big changeset. If this makes more sense as a separate repository - that also works for me.

### CODE REVIEW

**Comment by rth** [comment]
Thanks for working and contributing this back @gzuidhof ! Overall this looks great to me, though I'm certainly not an expert on modern web technologies. I'll add comments I have below, but we will need to find someone experienced on the frontend side for a review. 

Also,
 - @bcolloran  @wlach What would be the situation with respect to iodide given the end of active development status? Are we OK accepting something like this in pyodide which would potentially break compatibility with iodide for future releases (or require adaptation work which might or not happen)? 
 - @casatir WDTY with respect to Basthon integration?

**Comment by rth** [comment]
I think initially there was a desire to limit the number of technologies used in pyodide, to broaden the pool of potential contributors (as very few people would know all of these technologies) and keep the build easier. However this does solve a number real issues and assuming different pyodide modules are well isolated, so that one could contribute say to this package without caring about the rest maybe that would be OK.

In any case we would need to add a specific section to `docs/contributing.md` on how to build this. Thinking of Python developer who has no knowledge of JS tooling.

We would also need to build this in CI so the test suite actually uses the produced .js files. Probably by
 - adding bundler invocation in [Makefile](https://github.com/iodide-project/pyodide/blob/8c22c98a43e228c213f3f1dd7eae79ac88da4e90/Makefile#L114)
 - installing npm dependencies in [CI](https://github.com/iodide-project/pyodide/blob/8c22c98a43e228c213f3f1dd7eae79ac88da4e90/.circleci/config.yml#L34) (Node 8.9.1 is added to the PATH from emsdk). Though I guess we might also build it outside of the makefile with system nodejs which [should be](https://github.com/iodide-project/pyodide/blob/8c22c98a43e228c213f3f1dd7eae79ac88da4e90/Dockerfile#L9) more recent in the Docker image.
 - updating documentation on how to [build from sources](https://pyodide.readthedocs.io/en/latest/building_from_sources.html) 

I'm not sure what's the norm for frontend libraries, but it would have been great to have a test suite that checks some of these functions, independently of all the build system (or the presence of compiled packages). This could be another CI job, either in CircleCI or Github Actions.

> we can replace pyodide.js by browser.js and nothing should change from a user's perspective.

Do we need to change the name? I find the `browser.js` a bit surprising as a name for the pyodide loader and `pyodide.js` seems more explicit.

> The license of this package is MIT at the moment - but I guess it has to be MPL to be merged into here. 

Yes, I think keeping MPL would be preferable.

>  If this makes more sense as a separate repository - that also works for me.

It could make sense as a separate repo, except that it's very helpful to be able to run the test on everything in go. So I think we could start with it here, and move to a separate repo if necessary.

**Comment by gzuidhof** [comment]
Thank you for the comments, I agree with everything you said, I'll address the comments. Some notes:

* We can keep shipping `pyodide.js` and `webworker.js` in the pyodide bundle as it is currently done.
* I'm happy to use either CircleCI or Github Actions (I have experience with both). I'll stick with CircleCI for now as it seems to be where most of the building/testing happens.

* The reason it's called `browser.js` is because it makes more sense from a NPM perspective. When we publish this, one would be able to import it like:

```javascript
import {loadPyodide} from "pyodide-js"
```
Or in a script tag:
```html
<script src="https://unpkg.com/pyodide-js@0.15.0/browser.js"></script>
```
It's named after what kind of bundle it is, instead of what it is (as it's already inside the named NPM package anyway).

We can copy `src/pyodide-js/dist/browser.js` to `src/pyodide-js/pyodide.js` as they are (or should be!) equivalent.

* For testing one would usually default to [Jest](https://jestjs.io/), but I propose to leave that for a future PR. It does run its tests in Node though!

**Comment by bcolloran** [comment]
>     * @bcolloran  @wlach What would be the situation with respect to iodide given the end of active development status? Are we OK accepting something like this in pyodide which would potentially break compatibility with iodide for future releases (or require adaptation work which might or not happen)?

@rth, thanks for checking. Iodide will not be under active development for the foreseeable future, but we do not have any current plans to take down [Iodide.io](http://alpha.iodide.io), which continues to be actively used. Of course, much of this use includes pyodide, so ideally the _current version_ of pyodide would continue to function without interruption.

However, I don't think that _future_ versions of pyodide need to be constrained by Iodide compatibility -- Iodide's functionality is otherwise frozen at this time, so I think it's fine if the pyodide functionality available in Iodide is also frozen going forward.

So I guess I'd request that any current URLs from which pyodide is currently deployed should remain compatible with _any_ usage (not just Iodide) that expects a pyodide package that works the way pyodide works today, which might mean tweaking the build system or something to just never update the current deployment endpoints again if you decide to go in another direction.

Does this make sense and sound reasonable?

Also Roman, thanks for all of your amazing continued work on Pyodide!!! <3

**Comment by gzuidhof** [comment]
> installing npm dependencies in CI (Node 8.9.1 is added to the PATH from emsdk). Though I guess we might also build it outside of the makefile with system nodejs which should be more recent in the Docker image.

Alright, I think I managed to get the makefile right, the only issue now is the outdated Node version 8 (it's end of life anyway). Rollup requires at least Node >= 10, @rth could you maybe help me make the necessary changes in this PR?

Is it just a matter of changing the PATH to the system Node somehow? I'm a makefile beginner, so I don't know how it is currently setup.

**Comment by rth** [comment]
Thank you @bcolloran ! Yes, absolutely. We currently use versioned paths for download URLs, so even after next pyodide release, the last release of pyodide will work in idoide without modification.

@gzuidhof yes, I think it would be easier to do this outside of make (as part of CI) for now. At least in the docker image has Node v10.21.0. I'll try to look into it soon.

**Comment by rth** [comment]
@gzuidhof I updated the setup to run rollup outside of the Makefile. We can add it back once https://github.com/iodide-project/pyodide/pull/480 is merged which should also update the Nodejs version used in emsdk. For now running it separately should be fine.

I also merged master, following https://github.com/iodide-project/pyodide/pull/786 there was a bit of file reorganization in src/ 

So right now the build and the test suite should use the generated files. However there are still issues. Pyodide generally loads, but it seems that I'm not able to find `pyodide.runPython` function. For instance, when opening `build/console.html`, which includes pyodide.js as,
```
    <script type="text/javascript">
       window.languagePluginUrl = "./"
    </script>
    <script src="./pyodide.js"></script>
```
below is the error log I get 

<details>

```
ABI numbers differ. Expected 1, got 1 pyodide.js:289:27
Uncaught ABI numbers differ. Expected 1, got 1

Python path configuration: pyodide.asm.js:8:57160
  PYTHONHOME = '/' pyodide.asm.js:8:57160
  PYTHONPATH = (not set) pyodide.asm.js:8:57160
  program name = 'python3' pyodide.asm.js:8:57160
  isolated = 0 pyodide.asm.js:8:57160
  environment = 1 pyodide.asm.js:8:57160
  user site = 1 pyodide.asm.js:8:57160
  import site = 1 pyodide.asm.js:8:57160
  sys._base_executable = '' pyodide.asm.js:8:57160
  sys.base_prefix = '/' pyodide.asm.js:8:57160
  sys.base_exec_prefix = '/' pyodide.asm.js:8:57160
  sys.executable = '' pyodide.asm.js:8:57160
  sys.prefix = '/' pyodide.asm.js:8:57160
  sys.exec_prefix = '/' pyodide.asm.js:8:57160
  sys.path = [ pyodide.asm.js:8:57160
    'lib/python38.zip', pyodide.asm.js:8:57160
    '/lib/python3.8', pyodide.asm.js:8:57160
    '/lib/python3.8/lib-dynload', pyodide.asm.js:8:57160
  ] pyodide.asm.js:8:57160
Fatal Python error: init_fs_encoding: failed to get the Python codec of the filesystem encoding pyodide.asm.js:8:57160
Python runtime state: core initialized pyodide.asm.js:8:57160
ModuleNotFoundError: No module named 'encodings' pyodide.asm.js:8:57160
<empty string> pyodide.asm.js:8:57160
Current thread 0x00291854 (most recent call first): pyodide.asm.js:8:57160
<no Python frame> pyodide.asm.js:8:57160
Uncaught (in promise) TypeError: pyodide.runPython is not a function
```

</details>

Maybe I'm missing something. Could you have a look?

Also previously there were `pyodide.js` and `pyodide_dev.js` files, where the former loads from the CDN and the latter from the local build folder. I think it's not necessary and likely not compatible with this setup. So I removed `pyodide_dev.js` and `webworker_dev.js` in the hope that we could reproduce their functionality with by setting the `languagePluginUrl (or) pyodideArtifactsUrl` variable. WDYT?

Also had to rename `languageLoaderPlugin` to `languagePluginLoader` to avoid breaking backward compatibility. Not sure if the change was done on purpose. Generally I'm not against making the  names less specific from the iodide plugin system, but let's consider it separately from this PR so we can properly evalute all the bacward compatibility implications.

Another minor point is that when running `rollup` on Linux I get,
```
$ file dist/browser.js 
dist/browser.js: ASCII text, with CRLF, LF line terminators
```
which looks like incorrect line terminations to me (I see `^M` at the end of each line, when opened in a file editor). For comparison, the initial `pyodide.js` file is for instance,
```
$ file ../pyodide.js 
../pyodide.js: ASCII text
```
I haven't seen any encoding related setting in `rollup` so I'm not sure what this could be due to.

File: src/pyodide-js/src/pyodide.ts [hunk 1]
```diff
@@ -0,0 +1,292 @@
...
+      
+            for (const pkg in toLoad) {
+              let scriptSrc: string;
+              const packageUri = toLoad[pkg];
+              if (packageUri == 'default channel') {
+                scriptSrc = `${this.baseUrl}${pkg}.js`;
+              } else {
+                scriptSrc = `${packageUri}`;
+              }
+              // _messageCallback(`Loading ${pkg} from ${scriptSrc}`)
```
Comment (rth) [inline src/pyodide-js/src/pyodide.ts:162]: It kind of useful to have this information, at least in the simple `console.html`, as loading large packages can take a while..

File: src/pyodide-js/rollup.config.ts [hunk 2]
```diff
@@ -0,0 +1,42 @@
+import typescript from 'rollup-plugin-typescript2';
+import replace from '@rollup/plugin-replace';
+
+const pkg = require("./package.json");
+
+const pyodideCdnUrl = process.env["PYODIDE_CDN_URL"] || `https://cdn.jsdelivr.net/pyodide/${pkg.version}/full/`;
+const pyodideAbiNumber = process.env["PYODIDE_ABI_NUMBER"] === undefined ? `1` : process.env["PYODIDE_CDN_URL"];
```
Comment (rth) [inline src/pyodide-js/rollup.config.ts:7]:
```suggestion
const pyodideAbiNumber = process.env["PYODIDE_ABI_NUMBER"] === undefined ? `1` : process.env["PYODIDE_ABI_NUMBER"];
```
I think that would also fix the ABI version error.

Generally in terms of the name it would have been good to keep the same name as in the Makefile, `PYODIDE_PACKAGE_ABI` . Or else we need to change it quickly in the [FAQ](https://pyodide.readthedocs.io/en/latest/faq.html#how-to-detect-that-code-is-run-with-pyodide) before it gets carried over to various builds setups.

**Comment by gzuidhof** [comment]
> Also previously there were pyodide.js and pyodide_dev.js files, where the former loads from the CDN and the latter from the local build folder. I think it's not necessary and likely not compatible with this setup. So I removed pyodide_dev.js and webworker_dev.js in the hope that we could reproduce their functionality with by setting the languagePluginUrl (or) pyodideArtifactsUrl variable. WDYT?

Sounds good to me!

> Also had to rename languageLoaderPlugin to languagePluginLoader to avoid breaking backward compatibility. Not sure if the change was done on purpose.
Oops, that wasn't on purpose! I was hoping to find out these kinds of mistakes to show up when the tests would be ran.


> The files under pyodide-js/src also have correct line terminations. I haven't seen any encoding related setting in rollup so I'm not sure what this could be due to.
I have no idea here.. I pushed the PR from a windows machine, perhaps the source file has the wrong line endings?

File: src/pyodide-js/rollup.config.ts (hunk 2 above)
Comment (gzuidhof) [reply src/pyodide-js/rollup.config.ts:7]: Happy to change it to whatever you think makes sense :)

File: src/pyodide-js/src/pyodide.ts (hunk 1 above)
Comment (gzuidhof) [reply src/pyodide-js/src/pyodide.ts:162]: I agree, let's turn it back on. Worst case the end-user can filter it right now by string matching (which is what I'll do for Starboard probably). In the future we can think about more structured messages (some small object).

**Comment by rth** [comment]
So it looks like right now pyodide main package loads, but then it fails to load other python packages, at least according to the pytest results.

**Comment by gzuidhof** [comment]
It's a start I suppose. I'll have to find some time to set up the whole pipeline on my computer soon to debug it. Please don't wait for me to release 0.16.0, this can be in the version after it.

**Comment by rth** [comment]
Sounds good, thanks!

**Comment by dalcde** [comment]
What is the state of `pyodide.js` after this PR, it seems to me that we would be using the rollup bundle, but `pyodide.js` is not deleted in the PR.

**Comment by rth** [comment]
I have synced this PR with upstream/master to resolve conflicts, there is still quite a few changes that we need to take into account,
 - `PYODIDE_PACKAGE_ABI` was removed as is no longer used #991 
 - `PYODIDE_CDN_URL` is now `PYODIDE_BASE_URL` #855
 - Node JS 12 is now available as part of esdk https://github.com/iodide-project/pyodide/pull/480 suggesting that contributors need to install node is probably still better but just FYI
 - Once deployed we need to be able to use `webworker.js` both with the specified `PYODIDE_CDN_URL` and with `PYODIDE_CDN_URL='./'` in tests, which means we would need to set `languagePluginUrl` in the webworker.js at runtime https://github.com/iodide-project/pyodide/pull/855#issuecomment-747499425. I'm not sure if it's possible currently or what would be the best way to approach it.
 - @hoodmane made quite a few improvements to `pyodide.js` in https://github.com/iodide-project/pyodide/pull/998 I'm not sure how to best combine these two PRs.

>  it seems to me that we would be using the rollup bundle, but pyodide.js is not deleted in the PR.

 -  it should indeed be removed.


I'll push commits to address some of these points if you don't mind.

File: src/pyodide-js/package.json [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+  ],
+  "license": "MPL-2.0",
+  "bugs": {
+    "url": "https://github.com/project-iodide/pyodide/issues"
+  },
+  "homepage": "https://pyodide.readthedocs.io/",
+  "devDependencies": {
+    "@rollup/plugin-replace": "^2.3.4",
+    "@types/node": "^14.14.7",
+    "rimraf": "^3.0.2",
```
Comment (dalcde) [inline src/pyodide-js/package.json:28]: Given that we can only build on unix-like systems, can we use `rm -rf` instead?

File: src/pyodide-js/src/util.ts [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+  'globals',
+  'loadPackage',
+  'loadedPackages',
+  'pyimport',
+  'repr',
+  'runPython',
+  'runPythonAsync',
+  'version',
+  'autocomplete',
+];
```
Comment (hoodmane) [inline src/pyodide-js/src/util.ts:18]: This doesn't belong in this file. I'd say it's important enough to go in a file by itself, `api.ts` or something.

**Comment by hoodmane** [comment]
This looks good to me. I think we should manually redo this in several PRs so we have a better change history.

The biggest issue right now is that `conftest.py` is pretty broken. I've repeatedly had the problem that making a mild change to `pyodide.js` breaks all selenium tests with timeout exceptions even though manually doing the tests in the browser works. I will open an issue about selenium which I think we should consider to be a blocking for refactors to `pyodide.js`.

**Comment by rth** [comment]
> I think we should manually redo this in several PRs so we have a better change history.

Yes, maybe we could
 1. start by creating a separate PR that moves `pyodide.js`/`webworker.js` with minimal changes into a npm packages with the setup from this PR and check that CI is still fine.
 2. Split it into several JS files  as done in this PR
 3. Try to replace each file with the contents of this PR, also to have a more meaningful diff, and possibly integrated changes from #1084 etc.

WDYT?

**Comment by rth** [comment]
@gzuidhof Sorry in the end I'm closing this PR as it was too many changes to merge at once. 

We have however created a npm package (https://www.npmjs.com/package/pyodide, [src/js](https://github.com/pyodide/pyodide/tree/main/src/js)), with smaller iterative changes, some inspired by this PR. Additional improvements to this package, maybe adapted from this PR, would be very much welcome. As well as using typescript more https://github.com/pyodide/pyodide/issues/1729

**Comment by gzuidhof** [comment]
I completely understand, the NPM package kind of replaces this PR (which still needed some love), really happy to see that made it in :). It will make making small PR with improvements more feasible, which I am looking to contribute
